### ISSUE TITLE
`nonmininal_bool` appears to be confused about behaviour of the not operator for integers

### ISSUE THREAD

**Issue by zesterer** [rust-lang/rust-clippy#12384]
### Summary

I don't know much about the internals of clippy, but it seems like the `nonminimal_bool` lint blindly assumes that the operand of a not (`!`) operator is a boolean, leading it to suggest the following simplification:

`x == !y` => `x != y`

This is valid for booleans, but it is *not* valid for integers and changes the behaviour of the code. Worse still, conditions like this are often used to handle edge cases and the user unthinkingly applying the fix might have unsafe or dangerous results.

Clippy should probably first check the type of the operand: not just to handle the case of integers, but also because `Not` is an operator that can be overridden, allowing it to have arbitrary behaviour.

Example of this appearing in the wild:

```
error: this boolean expression can be simplified
    --> src/combinator.rs:1431:12
     |
1431 |         if self.at_most == !0 && self.at_least == 0 {
     |            ^^^^^^^^^^^^^^^^^^ help: try: `self.at_most != 0`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#nonminimal_bool
     = note: `-D clippy::nonminimal-bool` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(clippy::nonminimal_bool)]`
```

### Version

```text
The latest nightly
```

**Comment by Ethiraric** [comment]
This was fixed yesterday by #12372 
The following code does not generate a warning on `master`
```rs
fn main() {
    let x = 3;

    if x == !0 {
    }
}
```

**Comment by zesterer** [comment]
Great to hear, thanks!
